src/export/StructureDefinitionExporter.ts

```diff
@@ -637,30 +694,66 @@ export class StructureDefinitionExporter implements Fishable {
 
   applyDeferredRules() {
     this.deferredRules.forEach((rules, sd) => {
-      for (const rule of rules) {
+      for (const { rule, originalErr } of rules) {
+        let fishItem: string;
         if (typeof rule.value === 'string') {
-          const fishedValue = this.fishForFHIR(rule.value);
-          // An inline instance will fish up an InstanceDefinition, which can be used directly.
-          // Other instances will fish up an Object, which needs to be turned into an InstanceDefinition.
+          fishItem = rule.value;
+        } else if (
+          typeof rule.value === 'number' ||
+          typeof rule.value === 'bigint' ||
+          typeof rule.value === 'boolean'
+        ) {
+          fishItem = rule.rawValue;
+        }
+        let fishedValue = this.fishForFHIR(fishItem);
+        if (fishedValue) {
+          // If we are trying to assign a contained resource:
           // An InstanceDefinition of a resource needs a resourceType, so check for that property.
           // If we can't find a resourceType or an sdType, we have a non-resource Instance, which is no good
-          if (fishedValue) {
-            if (fishedValue instanceof InstanceDefinition && fishedValue.resourceType) {
-              try {
-                sd.setInstancePropertyByPath(rule.caretPath, fishedValue, this);
-              } catch (e) {
+          // if we can find the exported instance, it will have some useful metadata.
+          const instanceExporter = new InstanceExporter(this.tank, this.pkg, this.fisher);
+          const exportedInstance = instanceExporter.fishForFHIR(fishItem);
+          if (exportedInstance != null) {
+            fishedValue = exportedInstance;
+          }
+          const needsResourceType = /^contained(\[\d+\])?$/.test(rule.caretPath);
+          if (needsResourceType && fishedValue.resourceType == null) {
+            if (originalErr != null) {
+              logger.error(originalErr.message, rule.sourceInfo);
+            } else {
+              logger.error(`Could not find a resource named ${rule.value}`, rule.sourceInfo);
+            }
```

Comment:
What's this stuff with `contained`? Is this addressing a previously undiscovered bug?

Reply:
So, this is here in order to keep the test in `FHIRExporter.test.ts:139` working the same. The previous implementation here always needed something in `fishedValue.resourceType`, but now that assigning a non-resource Instance is allowed, I added this check on the path.

Thinking about it more, though, we can probably do without this check. It will mean that a different error is logged if a non-Instance resource is found when assigning to `contained`: specifically, it won't be the right type, since it isn't a descendant of `Resource`. But that's actually a fine thing to tell the author, unless I've overlooked something.